core/src/main/scala/org/apache/spark/status/ElementTrackingStore.scala

```diff
@@ -46,7 +50,26 @@ import org.apache.spark.util.kvstore._
  */
 private[spark] class ElementTrackingStore(store: KVStore, conf: SparkConf) extends KVStore {
 
-  private val triggers = new HashMap[Class[_], Seq[Trigger[_]]]()
+  private class LatchedTriggers(val triggers: Seq[Trigger[_]]) {
+    val countDeferred = new AtomicInteger(0)
+
+    def fireOnce(f: Seq[Trigger[_]] => Unit): Boolean = {
+      val shouldExecute = countDeferred.compareAndSet(0, 1)
+      if (shouldExecute) {
+        doAsync {
+          countDeferred.set(0)
+          f(triggers)
+        }
+      }
+      shouldExecute
```

Comment:
You could avoid the `WriteQueueResult` object with an inlined `if...else` here.

Reply:
I think there's a case where I needed the tri-value, but I'll take another peek.

Reply:
Hmm, my memory is faulty, this isn't a tri-value.  I'll inline here, sure.